### ISSUE TITLE
ghosting issue on n3dv dataset

### ISSUE THREAD

**Issue by Langwenchong** [donydchen/mvsplat#23]
Hello, I tested on the n3dv dataset using the model and weights provided by you and found that all NVS images have a ghosting issue. Since Gaussians are based on back-projecting to the world coordinate system from two views, I suspect the problem may be due to the accuracy of the camera parameters (as the camera parameters in the n3dv dataset are float64 and I converted them to float32) or could it be that image scaling is causing the Gaussian back-projection to not accurately align in depth (I scaled the images from 2k*2k to 512*512). Looking forward to your reply!
The following images are respectively the NVS RGB image  and the depth map from the source view：
<img src="https://github.com/donydchen/mvsplat/assets/58980433/0f753ea3-0f3e-4f10-abec-0161ccf33343" height="250px"/><img src="https://github.com/donydchen/mvsplat/assets/58980433/04567800-c8cc-4d01-8258-667128872495" height="250px"/><img src="https://github.com/donydchen/mvsplat/assets/58980433/41731bdf-db7d-4938-a500-a8b7033ba558"  height="250px" />

**Comment by JohnnyKimura** [comment]
您好，请问您的配置是什么呢？我用了40g 的A100 但是train 不了，我把batchsize改小了，他还是说我爆显存了，您有遇到这个问题嘛？或者您有什么思路吗，我是这方面的小白，刚起步什么都不会。
万分感谢您，打扰您了！
![图片](https://github.com/donydchen/mvsplat/assets/79315612/9dc1eb80-efd5-4884-a57c-721152368607)

**Comment by donydchen** [comment]
Hi @JohnnyKimura, I marked your comment off-topic above and addressed your issue in https://github.com/donydchen/mvsplat/issues/24; I hope you are okay with it. Cheers. 

----

Hi @Langwenchong, thanks for your interest in our work.

I have not tried applying MVSplat to N3DV, but the results you obtained do look a bit weird. I think the numerical differences between float32 and float16 might not be the cause since the differences should be slight. Scaling should be fine as long as the camera intrinsics you provided are normalized.

Below are my suggestions that might be helpful for your debugging:

* Double-check that the camera parameters align correctly with the code base, including both extrinsic and intrinsic. You can verify it by plotting the epipolar line between two input images (a.k.a context views); kindly refer to [this issue at pixelSplat](https://github.com/dcharatan/pixelsplat/issues/14) for more details.

* Double-check that the 'near' and 'far' values are set correctly. For more details, kindly refer to https://github.com/donydchen/mvsplat/issues/s.

* Try to scale the image to 256x256, as the released model is trained with 256x256 image resolution.

**Comment by Langwenchong** [comment]
I have another question to consult. After referring to your suggestions, I solved the issue of artifacts by realizing that the near and far planes of the input data need to be properly configured. However, it seems inconvenient to estimate the near and far planes offline every time for different scenes or viewpoints. Is it possible to implement an adaptive estimation of the near and far ranges for each viewpoint? While reviewing the `src/model/encoder/encoder_costvolume.py` in mvsplat,
![image](https://github.com/donydchen/mvsplat/assets/58980433/bc8a1c38-6129-402e-b3f5-7c239ff8a130)

 I noticed that you commented out the code from pixelsplat that recalculates the camera's near and far planes based on near_disparity and far_disparity. I tried using a similar approach as in pixelsplat by giving each viewpoint a large range of values based on `near_disparity=3.0` and `far_disparity=0.5` to build the cost volume, but the results were not good, leading to the reappearance of the ghosting issue. Do you have any good suggestions?

**Comment by donydchen** [comment]
Hi @Langwenchong, happy to see that you managed to relieve the previous ghosting effects.

In our experiments, we find that allowing the model to update the near and far range automatically makes the training a bit unstable, leading to less effective performance. That is the main reason why we removed that part of the codes.

If I remember correctly, the MVS community tends to set the near and far range of the cost volume as hyper-parameters, which can then be manually tuned during testing to get decent outputs for different scenes. Although it is a bit tedious, it is a reasonable solution as we can roughly calculate the depth range from the SfM reconstructed point cloud (typically coming with the COLMAP data).

Some suggestions I can think of for now:

* When training, set a larger value for the depth candidate of the cost volume (e.g., set `model.encoder.num_depth_candidates=256` in this project). This might allow the model to work on a larger depth range, hence more suitable for different scenes.
* Train / Finetune MVSplat on a more diverse dataset. The current released version of MVSplat is trained on RealEstate10K, which mainly contains real estate scenes. This makes it challenging for the trained model to generalize to other types of more complex in-the-wild scenes.

**Comment by haofeixu** [comment]
Hi @Langwenchong , thank you for your question. Is the background in your data pure white? If so, there might be some noises in the predicted depth of the white background, especially in the image boundaries, which accordingly leads to outliers in the reconstructed 3D Gaussians. It would be great if you could attach the input images, predicted depth of the input, and the reconstructed .ply file here, which would make it easier to figure out the issue, thanks.

**Comment by Langwenchong** [comment]
I apologize for the delayed response. Here is what I did: First, I used the open-source model weights you provided to fine-tune on the MVHumanNet dataset. Since the source images provided by MVHumanNet are in the following format, i.e., color images (though mostly black) + mask format, I made only a small adjustment during the fine-tuning process：
```python
 image = np.where(mask[..., np.newaxis] == 255.0, image, color)
```
That is, I kept the colors in the human region of the image and set the background entirely to black. This was then input into the model, which performed pixel-by-pixel back-projection with Gaussian encoding and decoding, and trained using color loss. Therefore, when performing final back-projection with Gaussian, the model also attempted to fit the background using black Gaussian. Since only color loss was used, the model's geometric reasoning for black Gaussian in the background might be abnormal (incorrect position, distribution, etc.) after convergence. However, the new viewpoint images inferred by the model were very good and had no issues. But when I checked the exported Gaussian, I found some purple Gaussian around the human body that shouldn't be there (these purple Gaussian did not appear in the new viewpoint images in the validation set output). This confuses me; what might be causing this issue and why does NVS rendered image not have purple color around human body?
Input 2 Context Images:
<img src="https://github.com/user-attachments/assets/734999d4-9c39-4392-af83-5e5c268af598" height="250px"> <img src="https://github.com/user-attachments/assets/8b420534-ac32-4742-ad08-135179fff54b" height="250px">
Depth map
<img src="https://github.com/user-attachments/assets/06fc1c58-d106-4a52-b7eb-0fab9b603668" height="250px"> <img src="https://github.com/user-attachments/assets/e491597e-c98b-4998-b030-9acd1a1f28b7" height="250px">
Renderd NVS Image(no purple gaussians):
<img src="https://github.com/user-attachments/assets/61ae9aa6-7314-45a4-84d7-45051f0122b2" height="250px"> <img src="https://github.com/user-attachments/assets/56c31976-128d-4d0b-97b8-0f3e47949be0" height="250px">

Output Gaussian .ply file saved in Goolge drive
https://drive.google.com/drive/folders/1AjCfKnOS7N23RpdGBD2OaDqRJxivutMd?usp=drive_link

Of course, I understand that the best approach should be to predict Gaussians only in the mask region and not generate Gaussians for the background pixels. Therefore, I am trying to modify the mvsplat code to construct the cost volume only in the human mask region. However, I have encountered some issues. This can lead to irregular cost volumes (possibly needing padding to fill edge areas), and mvsplat, similar to murf, directly constructs a volume storing the pixel matching costs and sampled pixel features at different depths. This can result in a very large cost volume. Even with 4x or 8x downsampling, when the input views are 1k or 2k, the memory usage can reach an astonishing 50-100GB, making it impossible to fit even one batch on a single A100 GPU.

In contrast, gps-gaussian first constructs a cost volume related only to depth prediction in the initial stage, estimates the depth map, and then in the second stage inputs the encoded depth map and source views into an MLP for decoding. Isn't this more reasonable? This is also why they can achieve inference with 1k input views on a single 4090 GPU.
